sel4-ftrace-merge/include/ftrace.h

```diff
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright 2023, Technology Innovation Institute
+ *
+ */
+#ifndef FTRACE_H
+#define FTRACE_H
+#include <stdint.h>
+#include <stddef.h>
+
+#define read_sized_data(T, d, f) do {		        \
+		int err;				\
+		d = malloc(sizeof(T));			\
+		assert(d);				\
+							\
+		err = fread(d, 1, sizeof(T), f);	\
+		assert(err == sizeof(T));		\
+							\
+		d = realloc(d, sizeof(T) + d->size);	\
+		assert(d);				\
+							\
+		err = fread(d->data, 1, d->size, f);	\
+		assert(err == d->size);			\
+	} while (0)
```

Comment:
Sorry my bad, just checked how Linux sources do this: move `do {` to its own line, with indentation of one tab. Then it should be perfect.

Reply:
Actually, I did grep `/src/linux` and can find pretty different cases... And pretty much with `do {` on the same line. But, I'm fine with any =)